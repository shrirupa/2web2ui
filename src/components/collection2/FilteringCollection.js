import React, { Component, Fragment } from 'react';
import { Panel, TextField } from '@sparkpost/matchbox';
import { Search } from '@sparkpost/matchbox-icons';
import PaginatingCollection from './PaginatingCollection';
import TableCollectionView from './TableCollectionView';
import _ from 'lodash';

const DEBOUNCE_FILTER_MS = 300;

const FilterBox = ({ onChange }) => <Panel sectioned>
  <TextField
    prefix={<Search />}
    onChange={(e) => onChange(e.target.value)}
  />
</Panel>;

export class FilteringCollection extends Component {
  state = {}

  static getDerivedStateFromProps(props, state) {
    if (state.rows) {
      return null;
    }

    // If we're given rows on mounth, use them
    return {
      rows: props.rows,
      rowCount: props.rowCount || (props.rows ? props.rows.length : 0),
      pattern: ''
    };
  }

  componentDidMount() {
    if (!this.state.rows) {
      this.handleFilterChange('');
    }
  }

  handleFilterChange = (pattern) => {
    this.setState({ rows: null, pattern });
    return Promise.resolve(this.props.fetchRows({ pattern }))
      .then(({ rows, rowCount }) => this.setState({ rows, rowCount }));
  };

  debouncedHandleFilterChange = _.debounce(this.handleFilterChange, DEBOUNCE_FILTER_MS);

  fetchFiltered = (filters) => this.props.fetchRows({
    pattern: this.state.pattern,
    ...filters
  })

  render() {
    const { collectionComponent: Coll = TableCollectionView } = this.props;
    const { rows, rowCount } = this.state;
    return <Fragment>
      <FilterBox onChange={this.debouncedHandleFilterChange} />
      {rows && <Coll {...this.props} rows={rows} rowCount={rowCount} fetchRows={this.fetchFiltered} />}
    </Fragment>;
  }
}

export const FilteringPaginatingCollection = ({ collectionComponent, ...rest }) => <FilteringCollection
  collectionComponent={(props) => <PaginatingCollection {...props} collectionComponent={collectionComponent} />}
  {...rest} />;
