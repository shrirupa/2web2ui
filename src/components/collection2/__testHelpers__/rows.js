export function mkRow(i = 0) {
  return { id: i, name: `row-name-${Math.floor(i / 5)}-${i}` };
}

export function mkRows(n) {
  const rows = [];
  for (let i = 0; i < n; i++) {
    rows.push(mkRow(i));
  }
  return rows;
}
