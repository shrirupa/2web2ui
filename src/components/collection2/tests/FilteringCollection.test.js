import React from 'react';
import { shallow } from 'enzyme';
import { FilteringCollection } from '../FilteringCollection';
import { mkRows } from '../__testHelpers__/rows';

describe('Component: FilteringCollection', () => {
  let props;
  let wrapper;

  beforeEach(() => {
    const rows = mkRows(10);
    props = {
      fetchRows: jest.fn().mockReturnValue({ rows, rowCount: rows.length }),
      rows
    };
    wrapper = shallow(<FilteringCollection {...props} />);
  });

  it('should render initial rows if provided', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should fetch rows on mount if none provided', () => {
    const { rows, ...rest } = props;
    wrapper = shallow(<FilteringCollection {...rest} />);
    expect(props.fetchRows).toHaveBeenCalledWith({ pattern: '' });
  });

  it('should render sync fetched rows', async () => {
    const { rows, ...rest } = props;
    wrapper = shallow(<FilteringCollection {...rest} />);
    await wrapper.instance().handleFilterChange();
    wrapper.update();
    expect(wrapper).toMatchSnapshot();
  });

  it('should render async fetched rows', async () => {
    const { rows, ...rest } = props;
    props.fetchRows.mockResolvedValue({ rows, rowCount: rows.length });
    wrapper = shallow(<FilteringCollection {...rest} />);
    await wrapper.instance().handleFilterChange();
    wrapper.update();
    expect(wrapper).toMatchSnapshot();
  });

  it('should fetch rows on filter change', () => {
    wrapper.instance().handleFilterChange('row');
    expect(props.fetchRows).toHaveBeenCalledWith({ pattern: 'row' });
  });

  it('should provide filtering for child collections', () => {
    wrapper.setState({ pattern: 'row' });
    const childFetchRows = wrapper.find('TableCollectionView').first().prop('fetchRows');
    childFetchRows({ currentPage: 2, perPage: 15 });
    expect(props.fetchRows).toHaveBeenCalledWith({
      pattern: 'row',
      currentPage: 2,
      perPage: 15
    });
  });
});
