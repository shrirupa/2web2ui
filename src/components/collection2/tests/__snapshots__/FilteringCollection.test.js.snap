// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: FilteringCollection should render async fetched rows 1`] = `
<React.Fragment>
  <FilterBox
    onChange={[Function]}
  />
  <TableCollectionView
    fetchRows={[Function]}
    rowCount={10}
    rows={
      Array [
        Object {
          "id": 0,
          "name": "row-name-0-0",
        },
        Object {
          "id": 1,
          "name": "row-name-0-1",
        },
        Object {
          "id": 2,
          "name": "row-name-0-2",
        },
        Object {
          "id": 3,
          "name": "row-name-0-3",
        },
        Object {
          "id": 4,
          "name": "row-name-0-4",
        },
        Object {
          "id": 5,
          "name": "row-name-1-5",
        },
        Object {
          "id": 6,
          "name": "row-name-1-6",
        },
        Object {
          "id": 7,
          "name": "row-name-1-7",
        },
        Object {
          "id": 8,
          "name": "row-name-1-8",
        },
        Object {
          "id": 9,
          "name": "row-name-1-9",
        },
      ]
    }
  />
</React.Fragment>
`;

exports[`Component: FilteringCollection should render initial rows if provided 1`] = `
<React.Fragment>
  <FilterBox
    onChange={[Function]}
  />
  <TableCollectionView
    fetchRows={[Function]}
    rowCount={10}
    rows={
      Array [
        Object {
          "id": 0,
          "name": "row-name-0-0",
        },
        Object {
          "id": 1,
          "name": "row-name-0-1",
        },
        Object {
          "id": 2,
          "name": "row-name-0-2",
        },
        Object {
          "id": 3,
          "name": "row-name-0-3",
        },
        Object {
          "id": 4,
          "name": "row-name-0-4",
        },
        Object {
          "id": 5,
          "name": "row-name-1-5",
        },
        Object {
          "id": 6,
          "name": "row-name-1-6",
        },
        Object {
          "id": 7,
          "name": "row-name-1-7",
        },
        Object {
          "id": 8,
          "name": "row-name-1-8",
        },
        Object {
          "id": 9,
          "name": "row-name-1-9",
        },
      ]
    }
  />
</React.Fragment>
`;

exports[`Component: FilteringCollection should render sync fetched rows 1`] = `
<React.Fragment>
  <FilterBox
    onChange={[Function]}
  />
  <TableCollectionView
    fetchRows={[Function]}
    rowCount={10}
    rows={
      Array [
        Object {
          "id": 0,
          "name": "row-name-0-0",
        },
        Object {
          "id": 1,
          "name": "row-name-0-1",
        },
        Object {
          "id": 2,
          "name": "row-name-0-2",
        },
        Object {
          "id": 3,
          "name": "row-name-0-3",
        },
        Object {
          "id": 4,
          "name": "row-name-0-4",
        },
        Object {
          "id": 5,
          "name": "row-name-1-5",
        },
        Object {
          "id": 6,
          "name": "row-name-1-6",
        },
        Object {
          "id": 7,
          "name": "row-name-1-7",
        },
        Object {
          "id": 8,
          "name": "row-name-1-8",
        },
        Object {
          "id": 9,
          "name": "row-name-1-9",
        },
      ]
    }
  />
</React.Fragment>
`;
